ipamd/ipamd_test.go

```diff
@@ -188,7 +192,108 @@ func TestIncreaseIPPool(t *testing.T) {
+		awsClient:     mockAWS,
+		k8sClient:     mockK8S,
+		networkClient: mockNetwork,
+		primaryIP:     make(map[string]string),
+	}
+
+	ds := datastore.NewDataStore()
+	ds.AddENI(secENIid, 1, false)
+	// Have to sleep for a minute in order for it to qualify as a deletable ENI because this interface
+	// is not exposed and not particularly testable...
+	time.Sleep(1 * time.Minute)
+	mockContext.dataStore = ds
+
+	mockAWS.EXPECT().FreeENI(secENIid)
+
+	mockContext.decreaseIPPool()
+}
+
+// Tests the mutex on the context to make sure that when a delete is called, a lock is
+// acquired on the context, so no add condition can cause a race
+func TestDecreaseIPPoolRaceCondition(t *testing.T) {
+	ctrl, mockAWS, mockK8S, mockNetwork := setup(t)
+	defer ctrl.Finish()
+
+	mockContext := &IPAMContext{
+		awsClient:     mockAWS,
+		k8sClient:     mockK8S,
+		networkClient: mockNetwork,
+		primaryIP:     make(map[string]string),
+	}
+
+	ds := datastore.NewDataStore()
+	ds.AddENI(secENIid, 1, false)
+	// Have to sleep for a minute in order for it to qualify as a deletable ENI because this interface
+	// is not exposed and not particularly testable...
+	time.Sleep(1 * time.Minute)
+	mockContext.dataStore = ds
+
+	// Ensures that the decrease and increase calls are called in the correct order
+	// This call is called by decreaseIPPool
+	first := mockAWS.EXPECT().FreeENI(secENIid).DoAndReturn(wait)
+
+	// These calls below are all called as a result of increaseIPPool
+	mockAWS.EXPECT().GetENILimit().Return(4, nil).After(first)
+	mockAWS.EXPECT().AllocENI().Return(secENIid, nil)
+	mockAWS.EXPECT().AllocAllIPAddress(secENIid)
+	mockAWS.EXPECT().GetAttachedENIs().Return([]awsutils.ENIMetadata{
+		awsutils.ENIMetadata{
+			ENIID:          primaryENIid,
+			MAC:            primaryMAC,
+			DeviceNumber:   primaryDevice,
+			SubnetIPv4CIDR: primarySubnet,
+			LocalIPv4s:     []string{ipaddr01, ipaddr02},
+		},
+		awsutils.ENIMetadata{
+			ENIID:          secENIid,
+			MAC:            secMAC,
+			DeviceNumber:   secDevice,
+			SubnetIPv4CIDR: secSubnet,
+			LocalIPv4s:     []string{ipaddr11, ipaddr12}},
+	}, nil)
+	mockAWS.EXPECT().GetPrimaryENI().Return(primaryENIid)
+	mockAWS.EXPECT().DescribeENI(secENIid).Return(
+		[]*ec2.NetworkInterfacePrivateIpAddress{
+			&ec2.NetworkInterfacePrivateIpAddress{
+				PrivateIpAddress: aws.String(ipaddr11), Primary: aws.Bool(false),
+			},
+			&ec2.NetworkInterfacePrivateIpAddress{
+				PrivateIpAddress: aws.String(ipaddr12), Primary: aws.Bool(false),
+			},
+		},
+		aws.String(testAttachmentID),
+		nil,
+	)
+	mockAWS.EXPECT().GetPrimaryENI().Return(primaryENIid)
+	mockNetwork.EXPECT().SetupENINetwork(gomock.Any(), secMAC, secDevice, secSubnet)
+
+	// Call decreaseIPPool and increaseIPPool in goroutines in quick succession and
+	// ensure that the decreaseIPPool calls are done first
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		mockContext.decreaseIPPool()
+		wg.Done()
+	}()
+	go func() {
+		mockContext.increaseIPPool()
```

Comment:
just to be safe you should make this one sleep a few milliseconds (maybe like 3-5) before calling, otherwise you might hit cases where the decrease starts first.

Reply:
Also would prob increase the sleep time just a tad so you can wait a bit longer here (who knows how fast the machine is thats running the tests).